### PR TITLE
Feat/gradient modes

### PR DESCRIPTION
**Pull request by TobiBu**

### CODE REVIEW

File: rubix/inference/optimize.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        trainable_params,
+        opt_state,
+        trainable_params,
+        jnp.array(jnp.inf),
+    )
+    (final_params, _, final_best_params, best_loss_val), (
+        loss_arr,
+        grad_norm_arr,
+        update_norm_arr,
+    ) = jax.lax.scan(scan_step, init_carry, None, length=max_steps)
```
[the remark this answers is not in the document]
Comment (TobiBu) [reply rubix/inference/optimize.py]: @copilot apply changes based on [this feedback](https://github.com/AstroAI-Lab/rubix/pull/206#discussion_r3110016315)

File: tests/test_inference_optimize.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        target=target,
+        learning_rate=0.1,
+        max_steps=250,
+        tol=1e-9,
+    )
+
+    assert result.loss_history[0] > result.loss_history[-1]
+    assert result.steps_run <= 250
+    assert result.best_loss <= result.loss_history[-1]
+
```
[the remark this answers is not in the document]
Comment (TobiBu) [reply tests/test_inference_optimize.py:72]: @copilot apply changes based on [this feedback](https://github.com/AstroAI-Lab/rubix/pull/206#discussion_r3110016252)

File: rubix/inference/optimize.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (TobiBu) [reply rubix/inference/optimize.py]: @copilot is this issue still present in the current code state?

File: tests/test_inference_optimize.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (TobiBu) [reply tests/test_inference_optimize.py:72]: @copilot is this still a relevant issue or is this now fixed?

File: rubix/inference/optimize.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+        grad_norm = optax.global_norm(grads)
+        update_norm = optax.global_norm(param_updates)
+
+        # Track best params on-device without Python-side branching
+        is_better = value < best_loss
+        new_best_loss = jnp.where(is_better, value, best_loss)
+        new_best_params = jax.tree_util.tree_map(
+            lambda new, old: jnp.where(is_better, new, old),
+            new_params,
```
[the remark this answers is not in the document]
Comment (TobiBu) [reply rubix/inference/optimize.py]: @copilot can you check again if this issue persist or if it was fixed in the latest commits?

File: tests/test_inference_optimize.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
```
[the remark this answers is not in the document]
Comment (TobiBu) [reply tests/test_inference_optimize.py:72]: @copilot apply changes based on [this feedback](https://github.com/AstroAI-Lab/rubix/pull/206#discussion_r3110395486)

File: tests/test_inference_optimize.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        max_steps=300,
+        tol=1e-9,
+        transforms=transforms,
+    )
+
+    final_age = result.params["stars"]["age"]
+    final_metallicity = result.params["stars"]["metallicity"]
+
+    assert result.converged
+    assert jnp.all(final_age > 0.0)
```
[the remark this answers is not in the document]
Comment (TobiBu) [reply tests/test_inference_optimize.py:120]: @copilot apply changes based on [this feedback](https://github.com/AstroAI-Lab/rubix/pull/206#discussion_r3110395512)

File: rubix/inference/optimize.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+    def scan_step(carry, _):
+        params, opt_state, best_params, best_loss, done = carry
+
+        def active_step(active_carry):
+            params, opt_state, best_params, best_loss, _ = active_carry
+            value, grads = jax.value_and_grad(train_loss)(params)
+            param_updates, new_opt_state = optimizer.update(grads, opt_state, params)
+            new_params = optax.apply_updates(params, param_updates)
+
```
[the remark this answers is not in the document]
Comment (TobiBu) [reply rubix/inference/optimize.py:132]: @copilot apply changes based on [this feedback](https://github.com/AstroAI-Lab/rubix/pull/206#discussion_r3110395539)

File: rubix/inference/optimize.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+            grad_norm = optax.global_norm(grads)
+            update_norm = optax.global_norm(param_updates)
```
[the remark this answers is not in the document]
Comment (TobiBu) [reply rubix/inference/optimize.py:134]: @copilot can you fix this based on this this feedback

[the remark this answers is not in the document]
Comment (TobiBu) [reply rubix/inference/optimize.py:134]: @copilot apply changes based on [this feedback](https://github.com/AstroAI-Lab/rubix/pull/206#discussion_r3111117219)
